### ISSUE TITLE
Parallel Routes cause redundant RSC fetches on navigation

### ISSUE THREAD

**Issue by steve-marmalade** [vercel/next.js#65878]
### Link to the code that reproduces this issue

https://github.com/marmalade-labs/parallel-route-rerender

### To Reproduce

1. Create production build of the application (`bun run build && bun start`) or navigate to https://parallel-route-rerender.vercel.app/a
2. Click on one of the cards in the bottom half of the page

### Current vs. Expected behavior

**Current Behavior**
If you open the Network tab in DevTools, when you click a card you will see 21 fetches for the current route (with different `?_rsc=` params), which I believe correspond to the 21 different parallel slots in the layout.

![image](https://github.com/vercel/next.js/assets/85196623/7e465ec1-47d3-4d2d-b83e-119f91d0f58a)

Maybe this is intended in the sense that each request might correspond to a specific slot. However, each request appears to be re-running the matched slots (which will always and only be `/app/[letter]/page.tsx` and `/app/[letter]/@grid/page.tsx`). I demonstrate by adding a 1 second sleep in those pages, but not in any others (which should just be rendering a no-op `default.tsx`), and yet we can see that each of the 21 requests takes > 1 second.

**Expected Behavior**
Whether or not we expect 21 different HTTP requests, I would only expect that each slot is rendered once.  So we should either see 1 request that takes ~ 1 second, or 21 requests where 2 of them take 1 second and the rest are <100 ms. 

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #1 SMP PREEMPT_DYNAMIC Tue, 07 May 2024 21:35:54 +0000
  Available memory (MB): 31886
  Available CPU cores: 8
Binaries:
  Node: 18.20.2
  npm: 10.8.0
  Yarn: 1.22.22
  pnpm: 8.15.5
Relevant Packages:
  next: 14.3.0-canary.68 // Latest available version is detected (14.3.0-canary.68).
  eslint-config-next: 14.2.3
  react: 18.3.1
  react-dom: 18.3.1
  typescript: 5.4.5
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Navigation, Parallel & Intercepting Routes

### Which stage(s) are affected? (Select all that apply)

next start (local), Vercel (Deployed)

**Comment by steve-marmalade** [comment]
@ztanner , tagging you in the hopes that I've misconfigured something in a way that will be easy for you to spot :pray:. This one is kind of a big deal for us because I believe it's increasing our server costs significantly (since we're doing the same work many times when a user navigates to a page).

**Comment by jimapl** [comment]
Do you do any api fetch requests on the server side?
I've noticed this behaviour on parallel routes where I rely on session and other fetched data. Where It builds the endpoint in dev, comes back with the data and rebuilds the parallel page. over and over again.

**Comment by ztanner** [comment]
Hi @steve-marmalade --

Thanks for the great repro and description, as always 😄 

As a quick workaround for the issue you're describing here, there're a few options:
- Add `prefetch={false}` to the links
- Or remove `app/[letter]/loading.tsx` and implement loading behavior with a regular <Suspense> boundary

Now, for an explanation of why it's happening: prefetch requests will return RSC data up to the nearest loading boundary. This is so that a prefetch doesn't lead to overfetching data, especially if you're linking to a page that has an expensive subtree. When a route is prefetched, the idea is that clicking the link will immediately show the prefetched loading boundary, while the part that suspended is streamed in. In this case, since there's a loading.tsx in the `[letter]` segment, upon navigation the router thinks that all of the parallel slots are missing data (because the server sent `null` data for them), triggering data fetches for them all.

I think I see a way to optimize this, and will work on a PR this coming week. Thanks for the heads up!

**Comment by steve-marmalade** [comment]
Thank you for the detailed description and the workaround (which does seem to alleviate the problem!)

Just want to clarify one thing for my own understanding:

> the router thinks that all of the parallel slots are missing data (because the server sent null data for them), triggering data fetches for them all.

Even in this case where the router triggers a data fetch for each slot, I still wouldn't expect those requests to take ~1 second each in my demo, because most of the slots are just returning `null` from `default.tsx`. Put another way, why does triggering a fetch for `/app/[letter]/@grid2/2/page.tsx` entail running `app/[letter]/page.tsx` and `app/[letter]/@grid/page.tsx`?

**Comment by steve-marmalade** [comment]
Heya @ztanner , I know there's a ton going on -- any chance this fix makes it into Canary this week?

**Comment by ztanner** [comment]
Hey @steve-marmalade -- I talked with some folks on the team and the concern is that, while many people use `default.tsx` to return `null`, you could just as easily have `default.tsx` return an expensive subtree, which would go against the reason we only return up to the loading segment for prefetches in the first place.

I'll need to think through if there's an alternate approach we could do here, which may take a little bit more time. 🙏

**Comment by steve-marmalade** [comment]
Got it, thanks for the update.

Relatedly: would you expect that scrolling to an `id` on the current page would also trigger the fetch requests for each slot? E.g. `<Link href="#section-a">Section A</Link>`. I actually didn't expect a Link with only a hash ID to make any fetch requests to the server. For now, I can replace with javascript `scrollIntoView`, but figured I'd raise in case it's unexpected to you as well.

**Comment by dalechyn** [comment]
> * Or remove `app/[letter]/loading.tsx` and implement loading behavior with a regular `<Suspense>` boundary

I tried implementing this approach by renaming all the `default`s, `page`s and `loading`s to `_default`s, `_page`s, `_loading`s and created custom wrapper everywhere with pretty much reexports of such wrapped in a `Suspense`.

Didn't help at all, still have lots of RSC fetches.

-----

> Add prefetch={false} to the links

Tried adding prefetch everywhere on top of it. Didn't help too.

I'm on the latest canary (`^15.0.0-canary.173`). This is really awful because those fetch calls are blocking and for users who are relatively far away, the load times multiplies by amount of redundant requests made.
![Screenshot 2024-09-29 at 23 12 03](https://github.com/user-attachments/assets/88174e47-03de-4c4d-accc-1bb3800638d8)

----- 
Downgrading to `@rc` didn't help too.

**Comment by dalechyn** [comment]
Hey @ztanner, this been hanging for a while, can you please check it out again?

I doubt that parallel routes are actually usable in NextJS unless this is fixed. And unfortunately for me, I nested lots of parallel routes with `loading.tsx` files expecting the feature to be stable.

**Comment by dalechyn** [comment]
I've debugged the issue (as much as I could), and was able to resolve it.

TLDR: I refactored my top-level parallel slots that were rendered on a client (via `use client`) to just be components wrapped with Suspense. I ended up with having 1 parallel slot (which is an async component that never gets to the client), and I have only one request as expected now.

https://github.com/user-attachments/assets/8d5db13e-f587-498d-9059-4901abd7dcb9

my dumpy dump file: https://gist.github.com/dalechyn/43da78d2dbdd1efc3c46fb4d1f04a91c

**Comment by steve-marmalade** [comment]
> Right now data is fetched per-page, rather than having the ability to more granularly fetch per-segment/slot. This is something we're planning to address soon so that the Next.js server can be smarter about only requesting the data it needs, rather than returning more data than necessary.
> 
> At the moment, when the router sees missing data for the `@grid3/default` segment, it triggers a fetch for the page it's currently on to retrieve that data (ie, `fetch("/a")`). The common layout won't be refetched, but everything in-between will. This means that it'll need to wait on any slower requests that also get matched in that request, since it's not currently possible to say "only fetch me the data for `/app/[letter]/@grid3/default.tsx`, but not `/app/[letter]/page.tsx`".

@ztanner , will the "Segment Cache" that you and @acdlite have been working on (e.g. https://github.com/vercel/next.js/pull/72875) provide the ability to do per-slot fetching for parallel routes (and therefore eliminate the redundant RSC fetches that were identified in this Issue)?

**Comment by marcoshernanz** [comment]
Thanks for the detailed repro. I retested it on the original Next.js 14 canary, Next.js 15.5.25, Next.js 16.3.4, and the current canary. The 21-request fan-out reproduces with the legacy cache in Next.js 14 and 15, but Next.js 16 makes one navigation request and renders each active slot once. Next.js 15 also has the fixed behavior with `experimental.clientSegmentCache` enabled.

The Segment Cache replaced this path in #72875, became the default in #84643, and the old prefetch cache was removed in #84977. Closing as fixed in Next.js 16. Please open a new issue with a current Next.js 16 repro if this still occurs.
